Revives Fortran case-sensitivity + full Fortran 2018 syntax highlighting

shahmoradi opened the pull request:
While Fortran is case-insensitive, modern-Fortran programmers (and most old-school Fortranners) rarely modify the case of Fortran entities throughout their codebase. So, if a decision is to be made on whether to keep case sensitivity, the readability of the generated documentation should be given more weight than the remote scenario where a Fortran developer uses mixed-case coding. In particular, there are ample Fortran codebases that use the CameCase style which becomes unreadable when Doxygen lowers the case of all letters. This PR of the Doxygen repository resolves this issue by removing all case-lowering instances of the Fortran entities in Doxygen. This effectively means that Doxygen treats Fortran code as case-sensitive, and Fortran users of Doxygen should respect case sensitivity in their codebase if fully-functional hyperlinking is desired. This does not affect other aspects of the documentation such as syntax highlighting and the functionalities related to the `iso_c_binding` module and C-Fortran interoperations. If any, it should only enhance C-Fortran interoperations as it interoperation rules defined by the Fortran standard are case-sensitive.

This PR also implements the following major enhancements to the Fortran syntax highlighting rules of Doxygen:

+   This PR implements a clean classification of (almost) all keywords in recent Fortran standards (2018/2008/2003/1995/1990/1977), fully consistent with the formal classification of keywords presented in [Fortran 2018 standard](https://j3-fortran.org/doc/year/18/18-007r1.pdf). Among the major categories are:
    1. Fortran attributes
    1. Fortran constructs
    1. Fortran statements
    1. Fortran intrinsic elemental functions
    1. Fortran intrinsic inquiry functions
    1. Fortran intrinsic transcendental functions
    1. Fortran intrinsic impure subroutines
    1. Fortran intrinsic pure subroutines
    1. Fortran intrinsic atomic subroutines
    1. Fortran intrinsic Coarray collective subroutines
    1. Fortran intrinsic IEEE modules and their entities
    1. Fortran intrinsic `iso_c_binding` module and its entities
    1. Fortran intrinsic `iso_fortran_env` module and its entities
    1. Fortran intrinsic specifiers (intrinsic-procedure argument names)
    1. Fortran intrinsic derived types

+ Over 600 new names of Fortran intrinsic functions, subroutines, modules, named constants, and specifiers (e.g., intrinsic procedure argument names) are now recognized and properly highlighted by Doxygen. The new set effectively covers all Fortran keywords in the latest Fortran standard 2018 and covers all keywords in 5 older standards F2008, F2003, F95, F90, F77.

+   Syntax recognition for MPI subroutines and entities, OpenMP directives, and Intel Fortran Compiler
    directives are now added.

+   The `real` and `logical` intrinsic functions are now properly recognized and highlighted as intrinsic functions or type declarations where appropriate.

+   Syntax recognition for Fortran `submodule`s and `module (function|subroutine|procedure)` now added.

+   Syntax recognition for Fortran intrinsic modules, including IEEE modules and their entities, `iso_fortran_enc` and `iso_c_binding` and their entities are now added.

+   Syntax recognition for new and modern Fortran constructs now added, for example, `block` and `associate` constructs.

+   Syntax recognition for all Fortran intrinsic Coarray parallel procedures and entities now added.

+   syntax recognition for new procedure attributes added (e.g., IMPURE, `(non_)?overridable`, ...).

+   The Fortran component separator symbol `%` is now recognized and highlighted. This is
    very useful for enhanced readability of Fortran codes as `%` is a rather busy symbol.
    
+   All Fortran intrinsic operators are now properly recognized and highlighted.

[Here is an example of Fortran documentation built with the modified Doxygen-Fortran implemented in this PR](https://cdslaborg.github.io/paramonte-kernel-doc/html/Array__mod_8f90_source.html).

shahmoradi committed:
Fortran case-sensitivity revived. Significant syntax highlighting enhancements.

+   Case-sensitivity of Fortran entities is now preserved:
    While it is true that Fortran is case-insensitive, modern-Fortran programmers (and even the old-schools)
    hardly modify the case of Fortran entities throughout their codebase. As such, if a decision is to be made
    on whether to keep case-sensitivity, the readability of the generated documentation should be given more weight
    than the remote scenario where a Fortran developer uses mixed-case programming. In particular, several Fortran
    codebases use CameCase style, which become unreadable when Doxygen lowers the case of all letters.
    This updated file removes all lowering of the cases of the Fortran entities.
    This effectively means that Doxygen treats Fortran code as text-sensitive,
    and Fortran users of Doxygen must respect the case-sensitivity.

+   Over 600 new names of Fortran intrinsic functions, subroutines, modules, named constants
    and specifiers (e.g., intrinsic procedure argument names) are now recognized and properly
    highlighted by Doxygen. The new set effectively covers all Fortran keywords in the latest
    Fortran standard 2018, and covers all keywords in 5 older standards F2008, F2003, F95, F90, F77.

+   Syntax recognition for MPI subroutines and entities, OpenMP directives, and Intel Fortran Compiler
    directives is now added.

+   The `real` and `logical` intrinsic functions are now properly recognized
    and highlighted as intrinsic functions or type declarations where appropriate.

+   Syntax recognition for Fortran `submodule`s and `module (function|subroutine|procedure) now added.

+   Syntax recognition for Fortran intrinsic modules, including IEEE modules and its entities,
    `iso_fortran_enc` and `iso_c_binding` and their entities is now added.

+   Syntax recognition for new and modern Fortran constructs now added,
    for example `block` and `associate` constructs.

+   Syntax recognition for all Fortran intrinsic Coarray parallel procedures and entities now added.

+   syntax recognition for new procedure attributes added (e.g., IMPURE, `(non_)?overridable`, ...).

+   The Fortran component separator symbol `%` is no recognized and highlighted. This is
    very useful for better readability of Fortran codes as `%` is a rather busy symbol.

+   All Fortran intrinsic operators are now properly recognized and highlighted.

shahmoradi committed:
Merge branch 'doxygen:master' into master

albert-github commented:
I didn't really do a review yet but I did a small test.

With the example: [example.tar.gz](https://github.com/doxygen/doxygen/files/7154547/example.tar.gz)
I.e.
**aa.f90**
```
!> \file


!> \brief The module
!> \details The module
module tst
contains
  !> \brief the first subroutine
  !> \details the first subroutine
  subroutine SUB1()
    call SUB2()
    call sub2()
    call Sub2()
  end subroutine Sub1
  !> \brief the second subroutine
  !> \details the second subroutine
  subroutine Sub2()
    write(*,*) "Hello"
  end subroutine SUB2
end module
```
**main.f90**
```
!> \file


!> \brief The main program
!> \details The main program
program pgm
  use tst
  call sub1()
end program
```
and *Doxyfile**
```
QUIET = YES
OPTIMIZE_FOR_FORTRAN = YES
CALL_GRAPH=YES
CALLER_GRAPH=YES
HAVE_DOT=YES
REFERENCED_BY_RELATION = YES
REFERENCES_RELATION    = YES
SOURCE_BROWSER = YES
STRIP_CODE_COMMENTS = NO
DOT_CLEANUP = NO
MAX_DOT_GRAPH_DEPTH = 1000
```

This gives in the 1.9.2 version:
![image](https://user-images.githubusercontent.com/5223533/133091603-e489a05a-49c3-42de-9389-b864f276a87f.png)

and

![image](https://user-images.githubusercontent.com/5223533/133091747-7125b11f-a330-42b4-8d2a-5286d1aa8a21.png)


With the patches as proposed in this pull request:

![image](https://user-images.githubusercontent.com/5223533/133091862-8ab4b313-acb7-4d36-b9cc-488e8b8687aa.png)

and

![image](https://user-images.githubusercontent.com/5223533/133091945-28ccb7bb-cd5c-4d06-8c1a-6a2276a090de.png)


(I did manipulate the colors of the links to make them clearer).

On directly sees:
- not all calls to sub1 are linked in the code (due to the inconsistent spelling but it is legal in Fortran and happens quite often in old and legacy code)
- the call / caller graphs are not correct

shahmoradi commented:
Thanks for testing. The broken links are not surprising to me (it would be the expected behavior with this PR). But I understand your concern and the necessity of supporting the language the way it is. Two alternatives that I see are the following,
1. Going the way you suggested in issue [#7868](https://github.com/doxygen/doxygen/issues/8768#issue-990644045) (Unfortunately, I do not have the knowledge of implementing it yet).
2. implementing an extra option for Doxygen to respect case sensitivity when requested by the user at runtime on the command line.

The latter seems to be easier to implement than the former, but again I do not currently have a deep understanding of the inner workings of Doxygen to judge what is possible? Let me know please if you believe any one of the above solutions should become an issue and I will open one for it, or if there are people already working on these or similar ideas that I could contribute.

The other parts of the PR (syntax highlighting) could be likely salvaged if the case-sensitivity is not desired. The two are easily separable.

albert-github commented:
A command line option is not really an alternative, it would be an entry in the doxygen settings file (`Doxyfile`) but I think this is still a workaround and it is better to really solve the problem and store the display name and the search name (as I indeed suggested before).

shahmoradi committed:
This commit resolves the doxygen bug related to not accepting public and private attributes on
separate independent lines in modules like:

    private
    public :: DistMulti_type

which would previously prevent Doxygen from generating depndency class heirarchy graphs. This bug is now resolved.

shahmoradi committed:
access attribute resolution

+ Resolves the module "public/private" object recognition issue by simply removing
all cases of private access recognition in Doxygen. Indeed, it is not really clear why such
private access recognition has been added to Doxygen in the first place as it seems entirely
unnecessary. The problem is that any public/private statement with followup object names,
messes with the objects' hyperlinking in the code, unless each object separately has its own
access attribute declared. Unless the attribute is public, the object hyperlinks to documentation
pages are disabled, which seems unnecessary.

+ Also resolves the `bind(C)` attribute highlighting issue.

shahmoradi committed:
submodule/proper specifier syntax recognition

shahmoradi committed:
minor indentation correction

shahmoradi committed:
Merge branch 'doxygen:master' into master

shahmoradi committed:
minor installation helper script

shahmoradi committed:
removing old Fortran comment ctyle Cc

albert-github commented on install.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+cd build && cmake -G "Unix Makefiles" .. && make || {
+    echo
+    echo "Doxygen build failed."
+    echo
+}
```
In my opinion this is script should not be in the repository, it is a private script used by the user.

doxygen replied:
I agree

rajeeja replied:
Is this good to go: on removal of the above script?

albert-github replied:
I don't think so see my comment about the old style comment.

albert-github commented on src/fortrancode.l:
```diff
@@ -932,14 +932,14 @@ INTRINSIC_DERIVED_TYPE          (C_FUNPTR|C_PTR|EVENT_TYPE|LOCK_TYPE|TEAM_TYPE)
                                                             endFontClass(yyscanner);
                                                         }
-
+ */
```
The old style comment is still standard Fortran.

rajeeja replied:
What is the fix for this?

albert-github replied:
Most  likely the not inclusion of the lines 935 and 942, see also https://github.com/doxygen/doxygen/pull/8774#issuecomment-918355501

shahmoradi replied:
The F77 comment style can be revived by removing the lines 935 and 942 as Albert says. However, adding it will mess up with Fortran90 source code lines that start with C or c. One option is to limit this comment style to .f and .for file extensions since this F77 comment style is not recognized and is illegal in the modern free-format Fortran files with extension .f90.
Another option is to modify the rule to better recognize the F77 comment-style where they appear without interfering with the syntax highlighting of free-format f90 source files.

albert-github replied:
@shahmoradi 
Can you give an example where normally the free formatted code is messed up due to this rule?
I can imagine that this is happening in the code coloring as I see in the lexical scanner (to determine the syntax) there is an attempt to convert from fixed formatted code to free formatted code, for source code output this conversion cannot be done and something different has to be done and is done as well but maybe not everywhere.

shahmoradi replied:
Yes, I do not have a minimal reproducible sample right now since I removed the F77 style comment rule in the fork longtime ago. But I remember this happened frequently in code sections within the documentation where Doxygen highlighted a `call xxx` statement starting at the first column as a comment line.

rajeeja commented:
@shahmoradi I tried to get the version from your repo: 
```
(base) jain @ jain ~/projects/doxygen/build (master)
└─ $ ▶ git remote -v
origin	git@github.com:cdslaborg/doxygen.git (fetch)
```

My doxyfile has some dot and graph features, but they end up being self loops, when in reality they are referenced. (different from one pictured below)
![image](https://user-images.githubusercontent.com/1466114/151680204-c5d94e91-d310-4961-a149-acf7880c912b.png)

shahmoradi committed:
enhancement: dot-graph artificial self-reference fixed.
Additionally, the syntax highlighting for multiple function attributes as prefixes now works properly.

shahmoradi commented:
> @shahmoradi I tried to get the version from your repo:
> 
> ```
> (base) jain @ jain ~/projects/doxygen/build (master)
> └─ $ ▶ git remote -v
> origin	git@github.com:cdslaborg/doxygen.git (fetch)
> ```
> 
> My doxyfile has some dot and graph features, but they end up being self loops, when in reality they are referenced. (different from one pictured below) ![image](https://user-images.githubusercontent.com/1466114/151680204-c5d94e91-d310-4961-a149-acf7880c912b.png)

@rajeeja Thanks for pointing this out. The graphs should look fine as long as case sensitivity is respected in the codebase. However, the dot-graph self-reference was a bug that I have now fixed in the lastest commit to the fork (not the oxygen main branch). The issue of submodules remains unresolved in this fork. 

Here is an example dot-graph generated with the latest commit:

![image](https://user-images.githubusercontent.com/4683401/151719907-0689cca7-5754-4dae-833f-d5c4a92332e6.png)

rajeeja commented:
cool, thanks for the fix.
 
The original version had caller graph (see pic below) also "Referenced by:" functions are missing, the version you have doesn't show caller graph: 

![image](https://user-images.githubusercontent.com/1466114/151736071-26b22f77-2a2b-475e-85b2-0e8053233cf4.png)

shahmoradi commented:
@rajeeja If you could post a minimal example that reproduces the bug, I'd appreciate it very much. Without the specific case, it is hard to figure out what is wrong.

rajeeja commented:
> @rajeeja If you could post a minimal example that reproduces the bug, I'd appreciate it very much. Without the specific case, it is hard to figure out what is wrong.
Understandable: How about this DOXYFILE: 
https://gist.github.com/rajeeja/5c785a17ef9ec59945f79ec9c16985b2

as I cannot share the private code, 
below is what the grep for the call of the function looks like (for the above picture):

./IO/IOMain/IO_output.F90:  !call IO_writeParticles.  If particles are not included in a simulation then
./IO/IOMain/IO_output.F90:  if(outputParticleFile) call IO_writeParticles( .false.)
./IO/IOMain/IO_writeCheckpoint.F90:  call IO_writeParticles( .true.)

shahmoradi commented:
@rajeeja Here is an example snapshot of the output by the most recent commit.

![image](https://user-images.githubusercontent.com/4683401/151739790-1122e702-9e72-4869-a408-556d754f6ca9.png)

I am not sure what is missing here. Perhaps you can elaborate further.

shahmoradi committed:
merging \fileinfo \lininfo improvements from Doxygen 1.9.5.
doxygen_lex.h is now added.

shahmoradi committed:
resolving the segfault in Fortran parser.

shahmoradi committed:
doxygen parser segfault still persists. this commit adds the valgrind analysis for further discussion with Doxygen team.

shahmoradi committed:
minor

shahmoradi committed:
minor

shahmoradi committed:
minor correction: revert back to 1.9.5 fix for double-free bug in fortran parser.
Although segfault errors still persist and point to a major underlying problem in Doxygen Fortran parser.

shahmoradi committed:
This commit changes the Fortran maximum line length from 72 to 10000.
The current fixed-format max length creates aesthetic problems in the example output files,
making anythign after column 72 appear like a comment.

shahmoradi committed:
Update README.md

shahmoradi committed:
Fortran lexer enhancements for parameterized derived types.

shahmoradi committed:
Merge branch 'master' of github.com:cdslaborg/doxygen
